### DISCUSSION TITLE
Difficulty Enabling HTTP3 with Traefik for Nextcloud AIO Container

### DISCUSSION THREAD

**Discussion by ollioddi** [nextcloud/all-in-one#4233] [Questions] [answered]
Hello,

I am encountering challenges enabling HTTP3 for a Nextcloud container using Traefik as reverse proxy. Despite configuring Traefik for HTTP3 support and forwarding UDP port 443 as recommended, HTTP3 remains non-functional.

**Configuration Details:**

- **Environment:** Running Nextcloud AIO version 7.12.1 on Ubuntu 22.04.
- **Traefik Setup:** Configured according to the guidance in PR #3965 on nextcloud/all-in-one, which documents enabling HTTP3 for Traefik.
- **Additional Context:** Noted Discussion #3461 suggesting the disabling of UDP exposure in proxy mode (now merged and active), and PR #2828 that seemingly supports exposing UDP for HTTP3.

**Issue Observed:**

Executing `netstat -lunp | grep 443` on the Nextcloud host indicates no services (including Apache) are listening on UDP port 443, contrary to expectations (or not).

**Concerns and Suggestions:**

The current documentation and configurations appear misaligned, causing confusion about enabling HTTP3 with Traefik for Nextcloud (and potentially others).

Therefore, i'd like to propose:
- **Documentation Improvement:** A clarification on the correct setup process.
- **Configuration Flexibility:** Consideration for setting HTTP3 support via environment variables (ENV) could offer more adaptability, especially given the evolving support for HTTP3 in reverse proxies. I think it was bold to merge https://github.com/nextcloud/all-in-one/discussions/3461 on the assumption that no known proxy can use http3? I apologize if i am mistaken though.

Thanks.

**Comment by szaimen** [comment]
@wuast94 could you maybe help here as apparently http3 does not work yet with traefik?

**Reply by ollioddi** [reply]
I think the only change necessary would be to modify https://github.com/nextcloud/all-in-one/pull/3515 to allow for exposing when in proxy mode. As mentioned, the container is simply not listening for HTTP3, due to being in proxy mode. I opened this discussion, because i dont have knowledge of the code base, but it seems "as simple as that". I think the default of being disabled is good for security. Perhaps enabling HTTP3 by exposing in proxy mode could be handled by ENV.

**Comment by simon-77** [comment, marked as answer]
> I am encountering challenges enabling HTTP3 for a Nextcloud container using Traefik as reverse proxy. Despite configuring Traefik for HTTP3 support and forwarding UDP port 443 as recommended, HTTP3 remains non-functional.

Hi @ollioddi 

I stumbled across this discussion while tweaking my Nextcloud & treafik setup. I just enabled HTTP3 for traefik and it immediately worked together with nextcloud - at least according to: https://http3check.net/

Just to be sure that you made all the neccessary changes, here is a list of things I had to change. This is of course dependent on the exact network setup you are using:

1. enable HTTP3 in treafik at your https "entrypoint" by uncommenting the `http3: {}` line in your traefik config https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md#traefik-2
NOTE: for Traefik 3 the `experimental: http3: true` is not required anymore

2. When running traefik in a docker container, open UDP port 443 alongside TCP port 443
3. When using port forwarding at your router, add an additional port forwarding for UDP port 443 alongside TCP port 443

Cheers Simon

**Reply by ollioddi** [reply]
Hi @simon-77. Thank you for replying with a fix for such an old thread.

I attempted enabling it again, and this time it _seems_ to work. I'm not sure if i forgot to expose it in docker compose or what.

Curiously https://http3check.net shows that HTTP3 is enabled for me. However, when i inspect the webui (Both Firefox and Chrome) and network requests **none of them** are using HTTP3. I generated an empty 5 GB file and tested uploading with it.

Furthermore, i run Nextcloud in a VM using AIO. If i run this command, it doesn't seem like anything is listening on UDP? I'm not sure if that's because HTTP3 starts as TCP and upgrades to UDP or what.

```
netstat -lunp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
udp        0      0 127.0.0.53:53           0.0.0.0:*                           98/systemd-resolved
udp        0      0 0.0.0.0:3478            0.0.0.0:*                           2079672/docker-prox
udp6       0      0 :::3478                 :::*                                2079681/docker-prox
```

It still feels like pseudo-http3 :-) I cannot really figure out if http3check.net detects http3 only between the client and traefik, and not client -> traefik -> Nextcloud

Have you done further research?

Edit: It also seems like a common misconception about HTTP3 that it should be enabled end-to-end, and not only between client-reverse proxy. So perhaps all is good.

**Reply by simon-77** [reply]
Great to hear that it worked for you this time.

Initially I also had the confirmation of http3check.net only, but my browser was not using it.

I was still fiddling around with my config and one issue was, that I had 2 different entrypoints in traefik (for external access and lan) and I only enabled it in one of the two. Also from what I remember, my browser did at some point not use HTTP3, but after an hour or so, I saw HTTP3 request in the development mode of the browser ... (was it some caching of headers by the browser, or what ever else ...)

To be sure that your browser supports HTTP3 (and your analysis of the usage of HTTP3 is correct), you can check this website:
https://cloudflare-quic.com/
